priv/static/phoenix_live_view.js

```diff
@@ -2104,6 +2109,87 @@ removing illegal node: "${(childNode.outerHTML || childNode.nodeValue).trim()}"
   };
 
   // js/phoenix_live_view/rendered.js
+  var VOID_TAGS = new Set([
+    "area",
+    "base",
+    "br",
+    "col",
+    "command",
+    "embed",
+    "hr",
+    "img",
+    "input",
+    "keygen",
+    "link",
+    "meta",
+    "param",
+    "source",
+    "track",
+    "wbr"
+  ]);
+  var endingTagNameChars = new Set([">", " ", "\n", "	", "\r"]);
+  var modifyRoot = (html, attrs, clearInnerHTML) => {
+    let i = 0;
+    let insideComment = false;
+    let insideTag = false;
+    let tag;
+    let beforeTagBuff = [];
+    while (i < html.length) {
+      let char = html.charAt(i);
+      if (insideComment) {
+        if (char === "-" && html.slice(i, i + 3) === "-->") {
+          insideComment = false;
+          beforeTagBuff.push("-->");
+          i += 3;
+        } else {
+          beforeTagBuff.push(char);
+          i++;
+        }
+      } else if (char === "<" && html.slice(i, i + 4) === "<!--") {
+        insideComment = true;
+        beforeTagBuff.push("<!--");
+        i += 4;
+      } else if (char === "<") {
+        insideTag = true;
+        let iAtOpen = i;
+        for (i; i < html.length; i++) {
+          if (endingTagNameChars.has(html.charAt(i))) {
+            break;
+          }
+        }
+        tag = html.slice(iAtOpen + 1, i);
+        break;
+      } else if (!insideComment && !insideTag) {
+        beforeTagBuff.push(char);
+        i++;
+      }
+    }
+    if (!tag) {
+      throw new Error(`malformed html ${html}`);
+    }
+    let attrsStr = Object.keys(attrs).map((attr) => attrs[attr] === true ? attr : `${attr}="${attrs[attr]}"`).join(" ");
+    let isVoid = VOID_TAGS.has(tag);
+    let closeTag = `</${tag}>`;
+    let newHTML;
+    let beforeTag = beforeTagBuff.join("");
```

Comment:
In JS, is it better to have a list and then join? or just always write to the string directly? Strings are mutable, right? so a string may be better.

Reply:
Or, even better, don't prepend to the list. Simply keep a counter and at the end beforeTag is `html.slice(0, beforeTagIndex)`.